src/mono/dbi/cordb-code.cpp

```diff
@@ -0,0 +1,125 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+// File: CORDB-CODE.CPP
+//
+
+#include <cordb-blocking-obj.h>
+#include <cordb-breakpoint.h>
+#include <cordb-code.h>
+#include <cordb-function.h>
+#include <cordb-process.h>
+#include <cordb.h>
+
+using namespace std;
+
+CordbCode::CordbCode(Connection* conn, CordbFunction* func) : CordbBaseMono(conn)
+{
+    this->m_pFunction = func;
+}
+
+HRESULT __stdcall CordbCode::IsIL(BOOL* pbIL)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbCode - IsIL - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT __stdcall CordbCode::GetFunction(ICorDebugFunction** ppFunction)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbCode - GetFunction - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT __stdcall CordbCode::GetAddress(CORDB_ADDRESS* pStart)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbCode - GetAddress - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT __stdcall CordbCode::GetSize(ULONG32* pcBytes)
+{
+    MdbgProtBuffer localbuf;
+    m_dbgprot_buffer_init(&localbuf, 128);
+
+    m_dbgprot_buffer_add_id(&localbuf, this->GetFunction()->GetDebuggerId());
+    int cmdId = conn->SendEvent(MDBGPROT_CMD_SET_METHOD, MDBGPROT_CMD_METHOD_GET_BODY, &localbuf);
+    m_dbgprot_buffer_free(&localbuf);
+
+    ReceivedReplyPacket* received_reply_packet = conn->GetReplyWithError(cmdId);
+    CHECK_ERROR_RETURN_FALSE(received_reply_packet);
+    MdbgProtBuffer* pReply = received_reply_packet->Buffer();
+
+    int code_size = m_dbgprot_decode_int(pReply->p, &pReply->p, pReply->end);
+    *pcBytes      = code_size;
+    LOG((LF_CORDB, LL_INFO1000000, "CordbCode - GetSize - IMPLEMENTED\n"));
+    return S_OK;
+}
+
+HRESULT __stdcall CordbCode::CreateBreakpoint(ULONG32 offset, ICorDebugFunctionBreakpoint** ppBreakpoint)
+{
+    // add it in a list to not recreate a already created breakpoint
+    CordbFunctionBreakpoint* bp = new CordbFunctionBreakpoint(conn, this, offset);
+    bp->QueryInterface(IID_ICorDebugFunctionBreakpoint, (void**)ppBreakpoint);
+    LOG((LF_CORDB, LL_INFO1000000, "CordbCode - CreateBreakpoint - IMPLEMENTED\n"));
+    return S_OK;
+}
+
+HRESULT __stdcall CordbCode::GetCode(
+    ULONG32 startOffset, ULONG32 endOffset, ULONG32 cBufferAlloc, BYTE buffer[], ULONG32* pcBufferSize)
+{
+    MdbgProtBuffer localbuf;
+    m_dbgprot_buffer_init(&localbuf, 128);
+
+    m_dbgprot_buffer_add_id(&localbuf, this->GetFunction()->GetDebuggerId());
+    int cmdId = conn->SendEvent(MDBGPROT_CMD_SET_METHOD, MDBGPROT_CMD_METHOD_GET_BODY, &localbuf);
+    m_dbgprot_buffer_free(&localbuf);
+
+    ReceivedReplyPacket* received_reply_packet = conn->GetReplyWithError(cmdId);
+    CHECK_ERROR_RETURN_FALSE(received_reply_packet);
+    MdbgProtBuffer* pReply = received_reply_packet->Buffer();
+
+    uint8_t* code = m_dbgprot_decode_byte_array(pReply->p, &pReply->p, pReply->end, (int32_t*)pcBufferSize);
+
+    memcpy(buffer, code, *pcBufferSize);
```

Comment:
It is possible that `buffer` could be `null`. In fact, it is typical that API of this sort are used with this pattern:

```
GetCode(a,b,c,null,&size);
BYTE* buffer = new BYTE[size];
GetCode(a,b,c,buffer,&size);
```
Therefore, if possible, we would like to reduce the cost for the first call, the code really only wants to know the size of the code so that it can allocate a buffer of proper size.

Also, it appears that we ignored the `startOffset` and `endOffset` parameters?

Reply:
In this case it's called GetSize before call GetCode, that is why the buffer always come with the right size. There is also no check of buffer in coreclr implementation: https://github.com/dotnet/runtime/blob/4f3545e2d1466512f53a2cd8b0db71667810c083/src/coreclr/debug/di/module.cpp#L3015
I implement the usage of startOffset and endOffset.